synapse/handlers/directory.py

```diff
@@ -307,15 +305,21 @@ def _update_canonical_alias(self, requester, user_id, room_id, room_alias):
             send_update = True
             content.pop("alias", "")
 
-        # Filter alt_aliases for the removed alias.
-        alt_aliases = content.pop("alt_aliases", None)
-        # If the aliases are not a list (or not found) do not attempt to modify
-        # the list.
-        if isinstance(alt_aliases, collections.Sequence):
-            send_update = True
-            alt_aliases = [alias for alias in alt_aliases if alias != alias_str]
-            if alt_aliases:
-                content["alt_aliases"] = alt_aliases
+        # Filter the alt_aliases property for the removed alias. Note that the
+        # value is not modified if alt_aliases is of an unexpected form.
+        try:
+            alt_aliases = content["alt_aliases"]
+            if isinstance(alt_aliases, (list, tuple)) and alias_str in alt_aliases:
+                send_update = True
+                alt_aliases = [alias for alias in alt_aliases if alias != alias_str]
+
+                if alt_aliases:
+                    content["alt_aliases"] = alt_aliases
+                else:
+                    content.pop("alt_aliases")
```

Comment:
```suggestion
                    del content["alt_aliases"]
```

might be marginally more efficient? not worth worrying about though.